### ISSUE TITLE
Descriptor `list[T]` becomes `list[Any]` when assigned a list of inferrable-type values

### ISSUE THREAD

**Issue by layday** [microsoft/pyright#1672]
**Describe the bug**
A descriptor `list[T]` where `T` is a concrete type becomes `list[Any]` when assigned a list of inferrable class-based values without the latter having being previously assigned to an intermediate variable.

**To Reproduce**
Type check the snippet below.

**Expected behavior**
That the last `bar.baz` would be `list[Foo]`.

**Screenshots or Code**
```python
# pyright: strict

from typing import Any, Generic, TypeVar

_T = TypeVar("_T")


class Baz(Generic[_T]):
    def __get__(self, instance: Any, owner: Any) -> _T: ...
    def __set__(self, instance: Any, value: _T) -> None: ...


class Foo:
    ...


class Bar:
    baz: Baz[list[Foo]]


bar = Bar()
reveal_type(bar.baz)   # Type of "bar.baz" is "list[Foo]"
foo_list = [Foo()]
reveal_type(foo_list)  # Type of "foo_list" is "list[Foo]"
bar.baz = foo_list
reveal_type(bar.baz)   # Type of "bar.baz" is "list[Foo]"
bar.baz = [Foo()]
reveal_type(bar.baz)   # Type of "bar.baz" is "list[Any]"

```

**VS Code extension or command-line**
1.1.123

**Comment by erictraut** [comment]
Thanks for the bug report. This will be fixed in the next release.

**Comment by erictraut** [comment]
This is addressed in pyright 1.1.124, which I just published. It will also be included in the next release of pylance.
